wiki/cpp/DEVELOPER_GUIDE.md

```diff
@@ -182,7 +182,58 @@ To avoid spurious code style violations we specify the exact clang-format versio
 Call CUDA APIs via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. These macros take care of checking the return values of the used API calls and generate an exception when the command is not successful. If you need to avoid an exception, e.g. inside a destructor, use `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVER_CHECK_NO_THROW ` (currently not available, see https://github.com/rapidsai/cuml/issues/229). These macros log the error but do not throw an exception.
 
 ## Logging
-Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+### Introduction
+Anything and everything about logging is defined inside [logger.hpp](../../cpp/include/cuml/common/logger.hpp). It uses [spdlog](https://github.com/gabime/spdlog) underneath, but this information is transparent to all.
+
+### Usage
+```cpp
+#include <cuml/common/logger.hpp>
+
+// Inside your method or function, use any of these macros
+CUML_LOG_TRACE("Hello %s!", "world");
+CUML_LOG_DEBUG("Hello %s!", "world");
+CUML_LOG_INFO("Hello %s!", "world");
+CUML_LOG_WARN("Hello %s!", "world");
+CUML_LOG_ERROR("Hello %s!", "world");
+CUML_LOG_CRITICAL("Hello %s!", "world");
+```
+
+### Changing logging level
+There are 7 logging levels with each successive level becoming quieter:
+1. CUML_LEVEL_TRACE
+2. CUML_LEVEL_DEBUG
+3. CUML_LEVEL_INFO
+4. CUML_LEVEL_WARN
+5. CUML_LEVEL_ERROR
+6. CUML_LEVEL_CRITICAL
+7. CUML_LEVEL_OFF
+Pass one of these as per your needs into the `setLevel()` method as follows:
+```cpp
+ML::Logger::get.setLevel(CUML_LEVEL_WARN);
+// From now onwards, this will print only WARN and above kind of messages
+```
+
+### Changing logging pattern
+Pass the [format string](https://github.com/gabime/spdlog/wiki/3.-Custom-formatting) as follows in order use a different logging pattern than the default.
+```cpp
+ML::Logger::get.setPattern(YourFavoriteFormat);
+```
+One can also use the corresponding `getPattern()` method to know the current format as well.
+
+### Temporarily changing the logging pattern
+Sometimes, we need to temporarily change the log pattern (eg: for reporting decision tree structure). This can be achieved in a RAII-like approach as follows:
+```cpp
+{
+  PatternSetter _(MyNewTempFormat);
+  // new log format is in effect from here onwards
+  doStuff();
+  // once the above temporary object goes out-of-scope, the old format will be restored
+}
+```
+
+### Tips
+* Do NOT end your logging messages with a newline! It is automatically added by spdlog.
+* The `CUML_LOG_DEBUG()` and `CUML_LOG_TRACE()` is by default not compiled due to the `CUML_ACTIVE_LEVEL` macro setup, for performance reasons. If you need them to be enabled, change this macro accordingly during compilation time
```

Comment:
Ah, I think i see why INFO was used in many places where I'd have used DEBUG. What do you think of compiling DEBUG by default but not TRACE? I suppose we could shift everything up a level and use INFO where we'd normally use DEBUG, but that seems like it might get confusing. 

We might also want to discuss, and agreement on, when a developer might want to use each log level, so that we can mention it here in a sentence or two. I see the INFO as being information I'd generally want a user to see when interested in the logs that's not relevant to debugging or troubleshooting a problem (so even when verbose=false). I see DEBUG as being important when a user sets `verbose=true` and TRACING really only important when a developer is trying to root cause a bug and needs to trace through a set of calls that might be happening in different processes (eg. tracing UCX messages).

Reply:
I think you have a good point. Enabling up until DEBUG level macros during compilation time certainly makes sense. Let me do that.

Reply:
Done.